eclair-core/src/main/scala/fr/acinq/eclair/blockchain/electrum/ElectrumWallet.scala

```diff
@@ -84,47 +86,141 @@ class ElectrumWallet(seed: BinaryData, client: ActorRef, params: ElectrumWallet.
     context.system.eventStream.publish(NewWalletReceiveAddress(data.currentReceiveAddress))
     data
   })
 
   when(DISCONNECTED) {
-    case Event(ElectrumClient.ElectrumReady(_, _), data) =>
+    case Event(ElectrumClient.ElectrumReady(_, _, _), data) =>
+      // subscribe to headers stream, server will reply with its current tip
       client ! ElectrumClient.HeaderSubscription(self)
       goto(WAITING_FOR_TIP) using data
   }
 
   when(WAITING_FOR_TIP) {
-    case Event(ElectrumClient.HeaderSubscriptionResponse(header), data) =>
+    case Event(ElectrumClient.HeaderSubscriptionResponse(height, header), data) if height < data.blockchain.height =>
+      log.info(s"elextrumx server is behind at ${height} we're at ${data.blockchain.height}, disconnecting")
+      sender ! PoisonPill
+      goto(DISCONNECTED) using data
+
+    case Event(ElectrumClient.HeaderSubscriptionResponse(height, header), data) if data.blockchain.bestchain.isEmpty =>
+      log.info("perfoming full sync")
+      // now ask for the first header after our latest checkpoint
+      client ! ElectrumClient.GetHeaders(data.blockchain.checkpoints.size * 2016, 2016)
+      // make sure there is not last ready message
+      goto(SYNCING) using data.copy(lastReadyMessage = None)
+
+    case Event(ElectrumClient.HeaderSubscriptionResponse(height, header), data) if header == data.blockchain.tip.header =>
       data.accountKeys.foreach(key => client ! ElectrumClient.ScriptHashSubscription(computeScriptHashFromPublicKey(key.publicKey), self))
       data.changeKeys.foreach(key => client ! ElectrumClient.ScriptHashSubscription(computeScriptHashFromPublicKey(key.publicKey), self))
-      // make sure there is not last ready message
-      goto(RUNNING) using data.copy(tip = header, lastReadyMessage = None)
+      goto(RUNNING) using notifyReady(data.copy(lastReadyMessage = None))
+
+    case Event(ElectrumClient.HeaderSubscriptionResponse(height, header), data) =>
+      client ! ElectrumClient.GetHeaders(data.blockchain.tip.height + 1, 2016)
+      log.info(s"syncing headers from ${data.blockchain.height} to ${height}")
+      goto(SYNCING) using data.copy(lastReadyMessage = None)
+
+    case Event(ElectrumClient.ElectrumDisconnected, data) =>
+      log.info(s"wallet got disconnected")
+      goto(DISCONNECTED) using data
+  }
+
+  when(SYNCING) {
+    case Event(ElectrumClient.GetHeadersResponse(start, headers, maxHeaders), data) if headers.isEmpty =>
+      // ok, we're all sync now
+      log.info(s"headers sync complete, tip=${data.blockchain.tip}")
+      data.accountKeys.foreach(key => client ! ElectrumClient.ScriptHashSubscription(computeScriptHashFromPublicKey(key.publicKey), self))
+      data.changeKeys.foreach(key => client ! ElectrumClient.ScriptHashSubscription(computeScriptHashFromPublicKey(key.publicKey), self))
+      goto(RUNNING) using notifyReady(data)
+
+    case Event(ElectrumClient.GetHeadersResponse(start, headers, maxHeaders), data) if start % 2016 == 0 =>
```

Comment:
I thought we stopped having multiple handers for the same type because it was too error prone?

Reply:
Will refactor